### PR TITLE
[python] allow plugins to set label in sortmethods

### PR DESCRIPTION
**Pull request by dagwieers**
## Description
The label is only used when no SortMethod was added to the listing, but as soon as a SortMethod is added the listing would suddenly be using the title (episode name).

This is problematic because we have listings that show episodes from different TV shows in the same listing (e.g. Recent items, Continue watching, Watch later or TV guides). Only showing the title (episode name) is useless in these listings, we tend to show the TV show name + the episode title.

It should always be using the label, just like when no SortMethod was added.

## Motivation and Context
The label is what the plugin developer wants to show explicitly, the title is the episode name. In a lot of listings add-ons need to be able to show the label, and not the episode title, without losing the ability to add SortMethods.

Add-on developers can influence the label, but ought to keep the metadata correct. Without this change add-on developers have to change the Episode name in the metadata in order to make listings look the way they want.

This fixes the issue (#17200)

## How Has This Been Tested?
- Tested using VRT NU Add-on in Kodi 19 on Linux x64

## Screenshots (if appropriate):
See #17156 

## Types of change
- [x] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [x] **Breaking change** (fix or feature that will cause existing functionality to change)
- [x] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

(For reviewers, the only change is replacing "%T" with "%L" so the Label
is used instead of the Title, see: https://github.com/xbmc/xbmc/compare/master...dagwieers:sortmethod-labels-simple)

## Checklist:
- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by dagwieers** [comment]
cc @ronie @mediaminister @michaelarnauts

File: xbmc/filesystem/PluginDirectory.cpp
Comment (Rechi) [inline xbmc/filesystem/PluginDirectory.cpp]: > The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

File: xbmc/filesystem/PluginDirectory.cpp
Comment (Rechi) [inline xbmc/filesystem/PluginDirectory.cpp]: > The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

File: xbmc/filesystem/PluginDirectory.cpp
Comment (Rechi) [inline xbmc/filesystem/PluginDirectory.cpp]: > The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

File: xbmc/filesystem/PluginDirectory.cpp
Comment (Rechi) [inline xbmc/filesystem/PluginDirectory.cpp]: > The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

File: xbmc/filesystem/PluginDirectory.cpp
Comment (Rechi) [inline xbmc/filesystem/PluginDirectory.cpp]: > The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

File: xbmc/filesystem/PluginDirectory.cpp
Comment (Rechi) [inline xbmc/filesystem/PluginDirectory.cpp]: > The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

File: xbmc/filesystem/PluginDirectory.cpp
Comment (Rechi) [inline xbmc/filesystem/PluginDirectory.cpp]: > The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

File: xbmc/filesystem/PluginDirectory.cpp
Comment (Rechi) [inline xbmc/filesystem/PluginDirectory.cpp]: > The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

File: xbmc/filesystem/PluginDirectory.cpp
Comment (Rechi) [inline xbmc/filesystem/PluginDirectory.cpp]: > The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

File: xbmc/filesystem/PluginDirectory.cpp
Comment (Rechi) [inline xbmc/filesystem/PluginDirectory.cpp]: > The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

File: xbmc/filesystem/PluginDirectory.cpp
Comment (Rechi) [inline xbmc/filesystem/PluginDirectory.cpp]: > Use spaces as tab policy with an indentation size of 2.

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#32-indentation

File: xbmc/filesystem/PluginDirectory.cpp
Comment (Rechi) [inline xbmc/filesystem/PluginDirectory.cpp]: > The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

**Review by Rechi — COMMENTED** [review]
Please follow the [current code guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md).

File: xbmc/filesystem/PluginDirectory.cpp
Comment (dagwieers) [reply xbmc/filesystem/PluginDirectory.cpp]: Alright, I didn't change the line length though, but I don't mind conforming to new rules.

**Comment by mediaminister** [comment]
I tested this with VRT NU Add-on in Kodi 19 on Linux x64 and it perfectly works.

File: xbmc/filesystem/PluginDirectory.cpp
Comment (dagwieers) [reply xbmc/filesystem/PluginDirectory.cpp]: I settled for what clang-format proposed to me. I hope that is fine.

Comment (Rechi) [reply xbmc/filesystem/PluginDirectory.cpp]: Your second commit for sure isn't the formatted code generated by clang-format, with the `.clang-format` configuration file from this repository.
Also there is no need for a separate code style commit, as every commit should follow the code guidelines, not only the endresult of a PR.

Comment (dagwieers) [reply xbmc/filesystem/PluginDirectory.cpp]: @Rechi Fine by me. It doesn't help reviewing the changes though.

Comment (dagwieers) [reply xbmc/filesystem/PluginDirectory.cpp]: My clang-format does not have an option to provide it with a config-file.

<details>
  <summary>Help output of my clang-format</summary>
[dag@moria xbmc]$ clang-format -h
OVERVIEW: A tool to format C/C++/Obj-C code.

If no arguments are specified, it formats the code from standard input
and writes the result to the standard output.
If <file>s are given, it reformats the files. If -i is specified
together with <file>s, the files are edited in-place. Otherwise, the
result is written to the standard output.

USAGE: clang-format [options] [<file> ...]

OPTIONS:
  -assume-filename=<string> - When reading from stdin, clang-format assumes this
                              filename to look for a style config file (with
                              -style=file).
  -cursor=<uint>            - The position of the cursor when invoking
                              clang-format from an editor integration
  -dump-config              - Dump configuration options to stdout and exit.
                              Can be used with -style option.
  -help                     - Display available options (-help-hidden for more)
  -i                        - Inplace edit <file>s, if specified.
  -length=<uint>            - Format a range of this length (in bytes).
                              Multiple ranges can be formatted by specifying
                              several -offset and -length pairs.
                              When only a single -offset is specified without
                              -length, clang-format will format up to the end
                              of the file.
                              Can only be used with one input file.
  -lines=<string>           - <start line>:<end line> - format a range of
                              lines (both 1-based).
                              Multiple ranges can be formatted by specifying
                              several -lines arguments.
                              Can't be used with -offset and -length.
                              Can only be used with one input file.
  -offset=<uint>            - Format a range starting at this byte offset.
                              Multiple ranges can be formatted by specifying
                              several -offset and -length pairs.
                              Can only be used with one input file.
  -output-replacements-xml  - Output replacements as XML.
  -style=<string>           - Coding style, currently supports:
                                LLVM, Google, Chromium, Mozilla, WebKit.
                              Use -style=file to load style configuration from
                              .clang-format file located in one of the parent
                              directories of the source file (or current
                              directory for stdin).
                              Use -style="{key: value, ...}" to set specific
                              parameters, e.g.:
                                -style="{BasedOnStyle: llvm, IndentWidth: 8}"
  -version                  - Display the version of this program
</details>

Comment (dagwieers) [reply xbmc/filesystem/PluginDirectory.cpp]: Ok, it does seem to pick up the `.clang-format` file, because it outputs the same content if I explicitly type: `clang-format -style=file xbmc/filesystem/PluginDirectory.cpp`.

Maybe we have a different clang-format?
```
[dag@moria xbmc]$ clang-format --version
LLVM (http://llvm.org/):
  LLVM version 3.4.2
  Optimized build.
  Built Nov  1 2018 (15:06:24).
  Default target: x86_64-redhat-linux-gnu
  Host CPU: x86-64
```

I have squashed the commits.

Comment (dagwieers) [reply xbmc/filesystem/PluginDirectory.cpp]: If I explicitly add `-length 100` I also get the same output.

Comment (dagwieers) [reply xbmc/filesystem/PluginDirectory.cpp]: I modified the changes to conform to the guidelines as good as possible.
- Indentation of 2 spaces
- curly brackets on newline

Anything else? Feel free to modify my branch which is probably more efficiënt.

File: xbmc/filesystem/PluginDirectory.cpp
Comment (ronie) [reply xbmc/filesystem/PluginDirectory.cpp]: this will result in the tracknumber being listed twice, once due %N and once as part of  %L

File: xbmc/filesystem/PluginDirectory.cpp
Comment (ronie) [reply xbmc/filesystem/PluginDirectory.cpp]: similar issue, the episodenumber is now duplicated.

**Comment by ronie** [comment]
frankly, this isn't really the way to go...
apart from the issues mentioned above, there are other:
- as Kodi uses %T in listings for local content, the would break the unified look&feel we have between local and plugin provided content
- every plugin ever released that uses sortmethods will now have to change their code to adopt this change

imo a proper implementation would be to add an option for plugins to overwrite the label, similar to how they can already overwrite label2.
see: https://github.com/xbmc/xbmc/blob/master/xbmc/interfaces/legacy/ModuleXbmcplugin.cpp#L66

**Comment by ronie** [comment]
untested: https://github.com/ronie/xbmc/commit/c024ec5026711e133fc32ff9f6f88aa7809046fb

File: xbmc/filesystem/PluginDirectory.cpp
Comment (dagwieers) [reply xbmc/filesystem/PluginDirectory.cpp]: So the Label is not what is provided by the developer when using `ListItem(label=<foobar>)`?
Now I am confused.

**Comment by dagwieers** [comment]
@mediaminister Can you test this patch from @ronie? I don't have a system where I can easily compile and run this code :-/

@ronie Would you like me to cherry-pick your changes to this branch, or do you prefer to create your own new PR for this? I am fine either way, just want to have a solution :-)

**Comment by mediaminister** [comment]
@dagwieers I had to completely rebuild the Kodi codebase, but I can confirm the sortlabel branch from @ronie works as expected.
If others want to test, I uploaded the modifications that VRT NU add-on needs: https://github.com/mediaminister/plugin.video.vrt.nu/tree/labelmask

**Comment by dagwieers** [comment]
I cherry-picked @ronie's changes into this PR. They are still authored/attributed to him.

**Review by ronie — APPROVED** [review]

**Comment by ronie** [comment]
thanx for testing it!

**Comment by phunkyfish** [comment]
jenkins build this please

File: xbmc/interfaces/legacy/ModuleXbmcplugin.h
Comment (enen92) [inline xbmc/interfaces/legacy/ModuleXbmcplugin.h]: Since v19 is a breaking change for the python API probably it makes sense to reorder the arguments. Or take this is consideration in https://github.com/xbmc/xbmc/pull/17456

Comment (ronie) [reply xbmc/interfaces/legacy/ModuleXbmcplugin.h]: yup, this is a typical example of the limitations of always trying to be backward compatible.
new options will have to be added in a clumsy / non-logical manner just to ensure we do not break anything.

**Comment by phunkyfish** [comment]
@dagwieers are you clear on how to progress this?

**Comment by dagwieers** [comment]
@phunkyfish I am waiting for this to get merged.

**Comment by enen92** [comment]
@dagwieers can you address my comment (reorder the function arguments)?

**Comment by dagwieers** [comment]
@enen92 It was not clear this was actually wanted, given that this was already approved, and no changes were requested in a review.

So we want to throw away backward compatibility in favor of cosmetics?

**Comment by enen92** [comment]
Backward compatibility was broken in matrix, so this is exactly the time to do cosmetic changes like this one

**Comment by dagwieers** [comment]
I was just notified about this PR needing work, but I have no clue what is expected from me.

Can one of the kodi-devs just update make the necessary change, and merge it?

It would be more efficient, especially when the needed changes are trivial and doesn't require any approval of the original author.

(Yes, if have *Allow edits by maintainers* enabled :wink: )

**Comment by mediaminister** [comment]
> it makes sense to reorder the arguments

I think you should change:
```c++
void addSortMethod(int handle, int sortMethod, const String& label2Mask = emptyString, const String& labelMask = emptyString);
```
to
```c++
void addSortMethod(int handle, int sortMethod, const String& labelMask = emptyString, const String& label2Mask = emptyString);
```

**Comment by dagwieers** [comment]
@mediaminister Ok, but my main point is that this PR could already have been merged for a triviality like this. In the projects I have maintained this was a policy we always applied: efficiency over process. If you can avoid another back-and-forth by merging it already with a trivial change, the original contributor is not really required anymore. Everyone wins.

**Comment by mediaminister** [comment]
I know, you're not alone: https://github.com/xbmc/xbmc/pull/17217

**Comment by dagwieers** [comment]
Please review, change (if needed) and merge this PR.

**Comment by enen92** [comment]
@dagwieers pr is already reviewed, I doubt anyone will rebase and work on the minimal changes on your behalf. As you are the main actor interested in this feature please either do the follow up or close the pull request. Worst case scenario someone might want to pick this up in the future and progress this.

**Comment by dagwieers** [comment]
@enen92 I have no clue what you are talking about as I already rebased this yesterday.

But please if something is ready to be merged except for some triviality, do fix the triviality and merge on the author's behalf, rather than wait for another roundtrip and/or needless delay.

**Comment by fuzzard** [comment]
It is what it is mate. If you want it merged, do the changes. If no one in the teams has an interest in a change, then no one is going to go about doing the changes for you. 

You can do the changes, or close the PR, up to you.

**Comment by dagwieers** [comment]
@fuzzard The change _has_ already been made.

**Comment by dagwieers** [comment]
I guess we are waiting for it to get stale again?

**Review by enen92 — APPROVED** [review]

**Comment by enen92** [comment]
@dagwieers sorry for the misunderstanding from my side, I thought the changes were still waiting to be made and you have just rebased. My apologies.
I think this is ready to go in.

**Comment by dagwieers** [comment]
Thanks everyone !

**Pull request merged**
